File: collector/destinations/mongodb.go
package destinations

import (
	"context"
	"fmt"

	"go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"

	"github.com/pcarpe4/bento/collector/core"
)

// mongodbDestination inserts each message's data as a document, with the
// collection metadata stored under "_meta".
type mongodbDestination struct {
	client     *mongo.Client
	collection *mongo.Collection
}

func init() {
	core.RegisterDestination("mongodb", newMongoDBDestination)
}

func newMongoDBDestination(cfg core.Fields) (core.Destination, error) {
	uri, err := cfg.RequiredString("url")
	if err != nil {
		return nil, err
	}
	database, err := cfg.RequiredString("database")
	if err != nil {
		return nil, err
	}
	collection, err := cfg.RequiredString("collection")
	if err != nil {
		return nil, err
	}

	opts := options.Client().ApplyURI(uri)
	if username := cfg.String("username", ""); username != "" {
		opts = opts.SetAuth(options.Credential{
			Username: username,
			Password: cfg.String("password", ""),
		})
	}

	client, err := mongo.Connect(opts)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to mongodb: %w", err)
	}

	return &mongodbDestination{
		client:     client,
		collection: client.Database(database).Collection(collection),
	}, nil
}

func (m *mongodbDestination) Write(ctx context.Context, batch []core.Message) error {
	docs := make([]any, 0, len(batch))
	for _, msg := range batch {
		doc := make(map[string]any, len(msg.Data)+1)
		for k, v := range msg.Data {
			doc[k] = v
		}
		doc["_meta"] = msg.Meta
		docs = append(docs, doc)
	}
	if _, err := m.collection.InsertMany(ctx, docs); err != nil {
		return fmt.Errorf("mongodb insert failed: %w", err)
	}
	return nil
}

func (m *mongodbDestination) Close(ctx context.Context) error {
	return m.client.Disconnect(ctx)
}

File: src/outputs/mongodb.rs
use anyhow::{bail, Context, Result};
use async_trait::async_trait;
use mongodb::bson::{self, Bson};
use serde::Deserialize;
use tokio::sync::OnceCell;

use crate::message::Message;
use crate::registry::Output;

/// Inserts each message's data as a document, with the collection metadata
/// stored under "_meta".
#[derive(Deserialize)]
struct MongoConfig {
    url: String,
    database: String,
    collection: String,
    #[serde(default)]
    username: String,
    #[serde(default)]
    password: String,
}

struct MongoDestination {
    cfg: MongoConfig,
    client: OnceCell<mongodb::Client>,
}

pub fn new(raw: &serde_yaml::Value) -> Result<Box<dyn Output>> {
    let cfg: MongoConfig = serde_yaml::from_value(raw.clone()).context("invalid config")?;
    for (value, field) in [
        (&cfg.url, "url"),
        (&cfg.database, "database"),
        (&cfg.collection, "collection"),
    ] {
        if value.is_empty() {
            bail!("config field {field:?} is required");
        }
    }
    Ok(Box::new(MongoDestination {
        cfg,
        client: OnceCell::new(),
    }))
}

impl MongoDestination {
    async fn client(&self) -> Result<&mongodb::Client> {
        self.client
            .get_or_try_init(|| async {
                let mut options = mongodb::options::ClientOptions::parse(&self.cfg.url)
                    .await
                    .context("failed to parse mongodb url")?;
                if !self.cfg.username.is_empty() {
                    options.credential = Some(
                        mongodb::options::Credential::builder()
                            .username(self.cfg.username.clone())
                            .password(self.cfg.password.clone())
                            .build(),
                    );
                }
                mongodb::Client::with_options(options).context("failed to create mongodb client")
            })
            .await
    }
}

#[async_trait]
impl Output for MongoDestination {
    async fn write(&self, batch: &[Message]) -> Result<()> {
        let mut docs = Vec::with_capacity(batch.len());
        for msg in batch {
            let mut doc = match bson::to_bson(&msg.data).context("failed to convert message")? {
                Bson::Document(doc) => doc,
                other => bson::doc! { "value": other },
            };
            doc.insert("_meta", bson::to_bson(&msg.meta)?);
            docs.push(doc);
        }

        self.client()
            .await?
            .database(&self.cfg.database)
            .collection::<bson::Document>(&self.cfg.collection)
            .insert_many(docs)
            .await
            .context("mongodb insert failed")?;
        Ok(())
    }

    async fn close(&self) -> Result<()> {
        if let Some(client) = self.client.get() {
            client.clone().shutdown().await;
        }
        Ok(())
    }
}
